File: dynamic-programming-hard/174.地下城游戏.cpp
/*
 * @lc app=leetcode.cn id=174 lang=cpp
 *
 * [174] 地下城游戏
 */
#include <bits/stdc++.h>
using namespace std;
// @lc code=start
class Solution {
public:
    int calculateMinimumHP(vector<vector<int>> &dungeon) {
        int m = dungeon.size(), n = dungeon[0].size();
        vector<vector<int>> dp(m, vector<int>(n, 0));
        for (int i = m - 1; i > -1; i--) {
            for (int j = n - 1; j > -1; j--) {
                if (i == m - 1 && j == n - 1)
                    dp[i][j] = max(1 - dungeon[i][j], 1);
                else if (i == m - 1 || j == n - 1)
                    dp[i][j] = max(dp[min(m - 1, i + 1)][min(n - 1, j + 1)] - dungeon[i][j], 1);
                else
                    dp[i][j] = max(min(dp[i + 1][j], dp[i][j + 1]) - dungeon[i][j], 1);
            }
        }
        return dp[0][0];
    }
};
// @lc code=end

File: binary-search-hard/174.地下城游戏.py
#
# @lc app=leetcode.cn id=174 lang=python3
#
# [174] 地下城游戏
#
import functools
from typing import List


# @lc code=start
class Solution:
    def calculateMinimumHP(self, dungeon: List[List[int]]) -> int:
        m, n = len(dungeon), len(dungeon[0])

        @functools.lru_cache(None)
        def dfs(x: int, y: int) -> int:
            if x == m - 1 and y == n - 1:
                return max(1-dungeon[x][y], 1)
            elif x == m - 1 or y == n - 1:
                return max(dfs(min(m-1, x+1), min(n-1, y+1))-dungeon[x][y], 1)
            return max(min(dfs(x+1, y), dfs(x, y+1)) - dungeon[x][y], 1)

        return dfs(0, 0)
# @lc code=end
